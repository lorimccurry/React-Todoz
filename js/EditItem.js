import React from 'react'

function EditItem (props) {
  const emptyEditID = ''
  return (
    <li key={props.id}>
      <input
        type='text'
        value={props.text}
        onChange={(e) => props.handleUpdateItemInput(e, props.id)}
        onKeyDown={(e) => props.onKeyDown(e, emptyEditID)}
        autoFocus
        name='text'
        className='update-input'
      />
      <button className='update-btn' onClick={() => props.handleEditID(emptyEditID)}>
        Update
      </button>
    </li>
  )
}

const { string, bool, func } = React.PropTypes
EditItem.propTypes = {
  text: string.isRequired,
  id: string.isRequired,
  completed: bool.isRequired,
  handleUpdateItemInput: func,
  onKeyDown: func,
  handleEditID: func
}

export default EditItem
