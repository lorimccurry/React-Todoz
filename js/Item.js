import React from 'react'

function Item (props) {
  return (
    <li>
      <input
        type='checkbox'
        checked={props.completed}
        onChange={(e) => props.handleUpdateItemInput(e, props.id)}
        onKeyDown={(e) => props.onKeyDown(e, props.id)}
        name='completed'
      />
      <p className='item-text'>{props.text}</p>
      <button onClick={() => props.handleEditID(props.id)} className='edit-btn'>Edit</button>
      <button onClick={() => props.handleDeleteItem(props.id)} className='delete-btn'>Delete</button>
    </li>
  )
}

const { string, bool, func } = React.PropTypes
Item.propTypes = {
  text: string.isRequired,
  id: string.isRequired,
  completed: bool.isRequired,
  handleEditID: func,
  handleDeleteItem: func,
  handleUpdateItemInput: func,
  onKeyDown: func
}

export default Item
