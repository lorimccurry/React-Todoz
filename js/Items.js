import React from 'react'
import Item from './Item'
import EditItem from './EditItem'

function Items (props) {
  const items = props.items
  if (!items) {
    return null
  }
  function renderListItems (items, props) {
    return items.map((item) => {
      if (props.editID === item.id) {
        return <EditItem
          key={item.id}
          {...item}
          handleEditID={(itemID) => props.handleEditID(itemID)}
          handleUpdateItemInput={(e, itemID) => props.handleUpdateItemInput(e, itemID)}
          onKeyDown={(e, itemID) => props.onKeyDown(e, itemID)}
        />
      } else {
        return <Item
          key={item.id}
          {...item}
          handleEditID={(itemID) => props.handleEditID(itemID)}
          handleUpdateItemInput={(e, itemID) => props.handleUpdateItemInput(e, itemID)}
          onKeyDown={(e, itemID) => props.onKeyDown(e, itemID)}
          handleDeleteItem={(itemID) => props.handleDeleteItem(itemID)}
        />
      }
    })
  }

  return (
    <ul>
      {renderListItems(items, props)}
    </ul>
  )
}

const { arrayOf, shape, string, bool, func } = React.PropTypes
Items.propTypes = {
  items: arrayOf(shape({
    text: string,
    id: string,
    completed: bool
  })),
  editID: string,
  handleDeleteItem: func,
  handleEditID: func,
  handleUpdateItemInput: func,
  onKeyDown: func
}

export default Items
