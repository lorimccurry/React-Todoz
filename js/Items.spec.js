import React from 'react'
import { shallow, mount } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import { spy } from 'sinon'
import Items from './Items'
import Item from './Item'
import EditItem from './EditItem'

describe('Items', () => {
  it('Items snapshot test', () => {
    const component = shallow(<Items />)
    const tree = shallowToJson(component)
    expect(tree).toMatchSnapshot()
  })

  it('should render zero items', () => {
    const component = shallow(<Items items={[]} />)
    expect(component.containsMatchingElement(Item)).toEqual(false)
  })

  it('should render undefined items', () => {
    const component = shallow(<Items items={undefined} />)
    expect(component.containsMatchingElement(Item)).toEqual(false)
  })

  it('should render some items', () => {
    const items = [
      {
        completed: false,
        id: 'askdjfo',
        text: 'sleep' 
      },
      {
        completed: false,
        id: '901ukcjv',
        text: 'eat yumz' 
      },
      {
        completed: false,
        id: 'iou12oiv',
        text: 'do the codez' 
      }
    ]
    const component = shallow(<Items items={items} />)
    expect(component.find(Item).length).toEqual(3)
  })

  it('should render 1 EditItem for item being edited', () => {
    const items = [
      {
        completed: false,
        id: 'askdjfo',
        text: 'sleep' 
      },
      {
        completed: false,
        id: '901ukcjv',
        text: 'eat yumz' 
      },
      {
        completed: false,
        id: 'iou12oiv',
        text: 'do the codez' 
      }
    ]
    const component = shallow(<Items
      items={items}
      editID={items[0].id}
      />)

    expect(component.find(Item).length).toEqual(2)
    expect(component.find(EditItem).length).toEqual(1)
  })

  it('should call handleEditID when edit button clicked', () => {
    const items = [
      {
        completed: false,
        id: 'askdjfo',
        text: 'sleep' 
      },
      {
        completed: false,
        id: '901ukcjv',
        text: 'eat yumz' 
      },
      {
        completed: false,
        id: 'iou12oiv',
        text: 'do the codez' 
      }
    ]
    const editItemSpy = spy()
    const component = mount(<Items
      items={items}
      handleEditID={editItemSpy}
      />)
    const editBtn = component.find('.edit-btn').first()
    editBtn.simulate('click')

    expect(editItemSpy.calledOnce).toEqual(true)
    expect(editItemSpy.calledWith('askdjfo')).toEqual(true)
  })

  it('should call handleDeleteItem when delete button clicked', () => {
    const items = [
      {
        completed: false,
        id: 'askdjfo',
        text: 'sleep' 
      },
      {
        completed: false,
        id: '901ukcjv',
        text: 'eat yumz' 
      },
      {
        completed: false,
        id: 'iou12oiv',
        text: 'do the codez' 
      }
    ]
    const deleteItemSpy = spy()
    const component = mount(<Items
      items={items}
      handleDeleteItem={deleteItemSpy}
      />)
    const deleteBtn = component.find('.delete-btn').first()
    deleteBtn.simulate('click')

    expect(deleteItemSpy.calledOnce).toEqual(true)
    expect(deleteItemSpy.calledWith('askdjfo')).toEqual(true)
  })

  it('should call handleUpdateItemInput when Item checkbox clicked', () => {
    const items = [
      {
        completed: false,
        id: 'askdjfo',
        text: 'sleep' 
      },
      {
        completed: false,
        id: '901ukcjv',
        text: 'eat yumz' 
      },
      {
        completed: false,
        id: 'iou12oiv',
        text: 'do the codez' 
      }
    ]
    const updateSpy = spy()
    const component = mount(<Items
      items={items}
      handleUpdateItemInput={updateSpy}
      />)
    const checkbox = component.find('input[type="checkbox"]').first()
    checkbox.simulate('change')

    expect(updateSpy.calledOnce).toEqual(true)
  })

  it('should call handleUpdateItemInput when EditItem text input changed', () => {
    const items = [
      {
        completed: false,
        id: 'askdjfo',
        text: 'sleep' 
      },
      {
        completed: false,
        id: '901ukcjv',
        text: 'eat yumz' 
      },
      {
        completed: false,
        id: 'iou12oiv',
        text: 'do the codez' 
      }
    ]
    const updateSpy = spy()
    const component = mount(<Items
      items={items}
      handleUpdateItemInput={updateSpy}
      editID={items[0].id}
      />)
    const input = component.find('input[type="text"]').first()
    input.simulate('change')

    expect(updateSpy.calledOnce).toEqual(true)
  })
})
