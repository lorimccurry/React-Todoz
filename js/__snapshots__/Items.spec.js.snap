// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Items Items snapshot test 1`] = `null`;
